### ISSUE TITLE
Wrong `missing-const-for-fn` suggestion for `&str`

### ISSUE THREAD

**Issue by frederikhors** [rust-lang/rust-clippy#14532]
### Summary

I have this rust code:

```rust
impl Something {
    pub fn action(&self) -> &str {
        &self.action
    }
}
```

clippy is suggesting this:

```
warning: this could be a `const fn`
   |
31 | /     pub fn action(&self) -> &str {
32 | |         &self.action
33 | |     }
   | |_____^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_const_for_fn
   = note: `-W clippy::missing-const-for-fn` implied by `-W clippy::nursery`
   = help: to override `-W clippy::nursery` add `#[allow(clippy::missing_const_for_fn)]`
help: make the function `const`
   |
31 |     pub const fn action(&self) -> &str {
   |         +++++
```

but if I add the `const` as clippy suggests i get this error:

```
error[E0015]: cannot perform non-const deref coercion on `std::string::String` in constant functions
     |
32   |         &self.action
     |         ^^^^^^^^^^^^
     |
     = note: attempting to deref into `str`
note: deref defined here
    --> C:\Users\Fred\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\alloc\src\string.rs:2686:5
     |
2686 |     type Target = str;
     |     ^^^^^^^^^^^
note: impl defined here, but it is not `const`
    --> C:\Users\Fred\.rustup\toolchains\stable-x86_64-pc-windows-msvc\lib/rustlib/src/rust\library\alloc\src\string.rs:2685:1
     |
2685 | impl ops::Deref for String {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^
     = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants
```

### Version

```text
rustc 1.86.0 (05f9846f8 2025-03-31)
binary: rustc
commit-hash: 05f9846f893b09a1be1fc8560e33fc3c815cfecb
commit-date: 2025-03-31
host: x86_64-pc-windows-msvc
release: 1.86.0
LLVM version: 19.1.7
```
